### ISSUE TITLE
snapshot delete getting stuck on 2.3

### ISSUE THREAD

**Issue by devarajphukan** [elastic/elasticsearch#18775]
**Elasticsearch version**: 2.3

**JVM version**: 1.8.0_91

**OS version**: Ubuntu 14.04

**Description of the problem including expected versus actual behavior**:

I am running a 8 node cluster, with 5 master eligible nodes. The cluster is configured to take snapshots to an azure container. When a snapshot process was taking too long I ran the delete command and the curl request got stuck. While the snapshot state shows "IN PROGRESS" and the some indices of the cluster shows "DONE" and some "ABORTED".
By running the delete command, an ongoing snapshot should have stopped and deleted the files. Now I can't delete or stop the snapshot process, nor can I start a new snapshot process.

**Steps to reproduce**:
1. Start a snapshot process.
2. Run Delete command while the snapshot process is running.
3. Try Starting a new Snapshot

**Provide logs (if relevant)**:

**curl -XGET "localhost:9200/_snapshot/my_backup/_all?pretty"**
{
  "snapshots" : [ {
    "snapshot" : "20160606-114625",
    "version_id" : 2030299,
    "version" : "2.3.2",
    "indices" : [ ".marvel-es-1-2016.06.06", "igdata_archive_v1", ".kibana", "fbdata_archive_v1", ".marvel-es-data-1", "twitterv1" ],
    "state" : "IN_PROGRESS",
    "start_time" : "2016-06-06T11:46:36.198Z",
    "start_time_in_millis" : 1465213596198,
    "failures" : [ ],
    "shards" : {
      "total" : 0,
      "failed" : 0,
      "successful" : 0
    }
  } ]
}

**curl -XGET "localhost:9200/_snapshot/my_backup/_status?pretty"**

{
  "snapshots" : [ {
    "snapshot" : "20160606-114625",
    "version_id" : 2030299,
    "version" : "2.3.2",
    "indices" : [ ".marvel-es-1-2016.06.06", "igdata_archive_v1", ".kibana", "fbdata_archive_v1", ".marvel-es-data-1", "twitterv1" ],
    "state" : "IN_PROGRESS",
    "start_time" : "2016-06-06T11:46:36.198Z",
    "start_time_in_millis" : 1465213596198,
    "failures" : [ ],
    "shards" : {
      "total" : 0,
      "failed" : 0,
      "successful" : 0
    }
  } ]
}
frrole@es-prod-0:~$ curl -XGET "localhost:9200/_snapshot/my_backup/_status?pretty"{
  "snapshots" : [ {
    "snapshot" : "20160606-114625",
    "repository" : "my_backup",
    "state" : "ABORTED",
    "shards_stats" : {
      "initializing" : 0,
      "started" : 0,
      "finalizing" : 0,
      "done" : 19,
      "failed" : 8,
      "total" : 27
    },
    "stats" : {
      "number_of_files" : 2569,
      "processed_files" : 1414,
      "total_size_in_bytes" : 894619893816,
      "processed_size_in_bytes" : 273115600981,
      "start_time_in_millis" : 1465213597658,
      "time_in_millis" : 2580664
    },
    "indices" : {
      "igdata_archive_v1" : {
        "shards_stats" : {
          "initializing" : 0,
          "started" : 0,
          "finalizing" : 0,
          "done" : 8,
          "failed" : 0,
          "total" : 8
        },
        "stats" : {
          "number_of_files" : 315,
          "processed_files" : 315,
          "total_size_in_bytes" : 640498637,
          "processed_size_in_bytes" : 640498637,
          "start_time_in_millis" : 1465213597710,
          "time_in_millis" : 17707
        },
        "shards" : {
          "0" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 42,
              "processed_files" : 42,
              "total_size_in_bytes" : 78022449,
              "processed_size_in_bytes" : 78022449,
              "start_time_in_millis" : 1465213597745,
              "time_in_millis" : 14713
            }
          },
          "1" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 49,
              "processed_files" : 49,
              "total_size_in_bytes" : 78835393,
              "processed_size_in_bytes" : 78835393,
              "start_time_in_millis" : 1465213597774,
              "time_in_millis" : 17218
            }
          },
          "2" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 32,
              "processed_files" : 32,
              "total_size_in_bytes" : 82103826,
              "processed_size_in_bytes" : 82103826,
              "start_time_in_millis" : 1465213597755,
              "time_in_millis" : 12653
            }
          },
          "3" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 61,
              "processed_files" : 61,
              "total_size_in_bytes" : 80134234,
              "processed_size_in_bytes" : 80134234,
              "start_time_in_millis" : 1465213597763,
              "time_in_millis" : 17654
            }
          },
          "4" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 29,
              "processed_files" : 29,
              "total_size_in_bytes" : 82384445,
              "processed_size_in_bytes" : 82384445,
              "start_time_in_millis" : 1465213598887,
              "time_in_millis" : 12094
            }
          },
          "5" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 36,
              "processed_files" : 36,
              "total_size_in_bytes" : 79855636,
              "processed_size_in_bytes" : 79855636,
              "start_time_in_millis" : 1465213597738,
              "time_in_millis" : 13382
            }
          },
          "6" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 33,
              "processed_files" : 33,
              "total_size_in_bytes" : 77978824,
              "processed_size_in_bytes" : 77978824,
              "start_time_in_millis" : 1465213597710,
              "time_in_millis" : 13688
            }
          },
          "7" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 33,
              "processed_files" : 33,
              "total_size_in_bytes" : 81183830,
              "processed_size_in_bytes" : 81183830,
              "start_time_in_millis" : 1465213597747,
              "time_in_millis" : 13045
            }
          }
        }
      },
      ".marvel-es-data-1" : {
        "shards_stats" : {
          "initializing" : 0,
          "started" : 0,
          "finalizing" : 0,
          "done" : 1,
          "failed" : 0,
          "total" : 1
        },
        "stats" : {
          "number_of_files" : 4,
          "processed_files" : 4,
          "total_size_in_bytes" : 5120,
          "processed_size_in_bytes" : 5120,
          "start_time_in_millis" : 1465213597655,
          "time_in_millis" : 1192
        },
        "shards" : {
          "0" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 4,
              "processed_files" : 4,
              "total_size_in_bytes" : 5120,
              "processed_size_in_bytes" : 5120,
              "start_time_in_millis" : 1465213597655,
              "time_in_millis" : 1192
            }
          }
        }
      },
      "fbdata_archive_v1" : {
        "shards_stats" : {
          "initializing" : 0,
          "started" : 0,
          "finalizing" : 0,
          "done" : 8,
          "failed" : 0,
          "total" : 8
        },
        "stats" : {
          "number_of_files" : 530,
          "processed_files" : 530,
          "total_size_in_bytes" : 1284798959,
          "processed_size_in_bytes" : 1284798959,
          "start_time_in_millis" : 1465213597772,
          "time_in_millis" : 68473
        },
        "shards" : {
          "0" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 83,
              "processed_files" : 83,
              "total_size_in_bytes" : 252055047,
              "processed_size_in_bytes" : 252055047,
              "start_time_in_millis" : 1465213611060,
              "time_in_millis" : 37487
            }
          },
          "2" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 38,
              "processed_files" : 38,
              "total_size_in_bytes" : 58202386,
              "processed_size_in_bytes" : 58202386,
              "start_time_in_millis" : 1465213597774,
              "time_in_millis" : 13730
            }
          },
          "1" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 49,
              "processed_files" : 49,
              "total_size_in_bytes" : 160284115,
              "processed_size_in_bytes" : 160284115,
              "start_time_in_millis" : 1465213597781,
              "time_in_millis" : 22071
            }
          },
          "4" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 49,
              "processed_files" : 49,
              "total_size_in_bytes" : 85370756,
              "processed_size_in_bytes" : 85370756,
              "start_time_in_millis" : 1465213612519,
              "time_in_millis" : 15931
            }
          },
          "3" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 78,
              "processed_files" : 78,
              "total_size_in_bytes" : 165212008,
              "processed_size_in_bytes" : 165212008,
              "start_time_in_millis" : 1465213597772,
              "time_in_millis" : 25008
            }
          },
          "6" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 67,
              "processed_files" : 67,
              "total_size_in_bytes" : 129844814,
              "processed_size_in_bytes" : 129844814,
              "start_time_in_millis" : 1465213597786,
              "time_in_millis" : 20858
            }
          },
          "5" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 64,
              "processed_files" : 64,
              "total_size_in_bytes" : 117159311,
              "processed_size_in_bytes" : 117159311,
              "start_time_in_millis" : 1465213615057,
              "time_in_millis" : 19244
            }
          },
          "7" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 102,
              "processed_files" : 102,
              "total_size_in_bytes" : 316670522,
              "processed_size_in_bytes" : 316670522,
              "start_time_in_millis" : 1465213615536,
              "time_in_millis" : 50709
            }
          }
        }
      },
      ".kibana" : {
        "shards_stats" : {
          "initializing" : 0,
          "started" : 0,
          "finalizing" : 0,
          "done" : 1,
          "failed" : 0,
          "total" : 1
        },
        "stats" : {
          "number_of_files" : 11,
          "processed_files" : 11,
          "total_size_in_bytes" : 51305,
          "processed_size_in_bytes" : 51305,
          "start_time_in_millis" : 1465213597658,
          "time_in_millis" : 2041
        },
        "shards" : {
          "0" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 11,
              "processed_files" : 11,
              "total_size_in_bytes" : 51305,
              "processed_size_in_bytes" : 51305,
              "start_time_in_millis" : 1465213597658,
              "time_in_millis" : 2041
            }
          }
        }
      },
      ".marvel-es-1-2016.06.06" : {
        "shards_stats" : {
          "initializing" : 0,
          "started" : 0,
          "finalizing" : 0,
          "done" : 1,
          "failed" : 0,
          "total" : 1
        },
        "stats" : {
          "number_of_files" : 22,
          "processed_files" : 22,
          "total_size_in_bytes" : 1381698,
          "processed_size_in_bytes" : 1381698,
          "start_time_in_millis" : 1465213611530,
          "time_in_millis" : 3997
        },
        "shards" : {
          "0" : {
            "stage" : "DONE",
            "stats" : {
              "number_of_files" : 22,
              "processed_files" : 22,
              "total_size_in_bytes" : 1381698,
              "processed_size_in_bytes" : 1381698,
              "start_time_in_millis" : 1465213611530,
              "time_in_millis" : 3997
            }
          }
        }
      },
      "twitterv1" : {
        "shards_stats" : {
          "initializing" : 0,
          "started" : 0,
          "finalizing" : 0,
          "done" : 0,
          "failed" : 8,
          "total" : 8
        },
        "stats" : {
          "number_of_files" : 1687,
          "processed_files" : 532,
          "total_size_in_bytes" : 892693158097,
          "processed_size_in_bytes" : 271188865262,
          "start_time_in_millis" : 1465213597920,
          "time_in_millis" : 2580402
        },
        "shards" : {
          "6" : {
            "stage" : "FAILURE",
            "stats" : {
              "number_of_files" : 221,
              "processed_files" : 65,
              "total_size_in_bytes" : 124216528136,
              "processed_size_in_bytes" : 41597344471,
              "start_time_in_millis" : 1465213597987,
              "time_in_millis" : 2580283
            },
            "reason" : "IndexShardSnapshotFailedException[Aborted]"
          },
          "7" : {
            "stage" : "FAILURE",
            "stats" : {
              "number_of_files" : 218,
              "processed_files" : 69,
              "total_size_in_bytes" : 127543286629,
              "processed_size_in_bytes" : 38811410641,
              "start_time_in_millis" : 1465213610744,
              "time_in_millis" : 2567483
            },
            "reason" : "IndexShardSnapshotFailedException[Aborted]"
          },
          "2" : {
            "stage" : "FAILURE",
            "stats" : {
              "number_of_files" : 227,
              "processed_files" : 74,
              "total_size_in_bytes" : 127390735566,
              "processed_size_in_bytes" : 36220956572,
              "start_time_in_millis" : 1465213597877,
              "time_in_millis" : 2580272
            },
            "reason" : "IndexShardSnapshotFailedException[Aborted]"
          },
          "3" : {
            "stage" : "FAILURE",
            "stats" : {
              "number_of_files" : 0,
              "processed_files" : 0,
              "total_size_in_bytes" : 0,
              "processed_size_in_bytes" : 0,
              "start_time_in_millis" : 0,
              "time_in_millis" : 0
            }
          },
          "4" : {
            "stage" : "FAILURE",
            "stats" : {
              "number_of_files" : 245,
              "processed_files" : 111,
              "total_size_in_bytes" : 125007807167,
              "processed_size_in_bytes" : 41813102390,
              "start_time_in_millis" : 1465213597920,
              "time_in_millis" : 2580319
            },
            "reason" : "IndexShardSnapshotFailedException[Aborted]"
          },
          "5" : {
            "stage" : "FAILURE",
            "stats" : {
              "number_of_files" : 265,
              "processed_files" : 65,
              "total_size_in_bytes" : 134373968301,
              "processed_size_in_bytes" : 39058558697,
              "start_time_in_millis" : 1465213598117,
              "time_in_millis" : 2580205
            },
            "reason" : "IndexShardSnapshotFailedException[Aborted]"
          },
          "0" : {
            "stage" : "FAILURE",
            "stats" : {
              "number_of_files" : 256,
              "processed_files" : 86,
              "total_size_in_bytes" : 130201274206,
              "processed_size_in_bytes" : 39822216979,
              "start_time_in_millis" : 1465213598039,
              "time_in_millis" : 2580168
            },
            "reason" : "IndexShardSnapshotFailedException[Aborted]"
          },
          "1" : {
            "stage" : "FAILURE",
            "stats" : {
              "number_of_files" : 255,
              "processed_files" : 62,
              "total_size_in_bytes" : 123959558092,
              "processed_size_in_bytes" : 33865275512,
              "start_time_in_millis" : 1465213628544,
              "time_in_millis" : 2549621
            },
            "reason" : "IndexShardSnapshotFailedException[Aborted]"
          }
        }
      }
    }
  } ]
}

**Comment by clintongormley** [comment]
@imotov @dadoonet any ideas what might be happening here?

**Comment by imotov** [comment]
@devarajphukan did you wait for the delete operation to finish before starting a new snapshot?

**Comment by devarajphukan** [comment]
@imotov, the curl request for delete was still hanging on when I tried a new snapshot process. Had to do a rolling restart of the whole cluster to stop the delete request and start a new snapshot. Correct me if I am wrong but from the documentation what I understood was we can make a delete request on an ongoing snapshot process to stop it. 
Thanks.

**Comment by imotov** [comment]
@devarajphukan yes, the delete request was in the process of cancelling the running snapshot but got stuck for some reason. How long was the delete operation hanging before you tried to start a new snapshot? Did you have restart the whole cluster? I would expect just restarting the current master should have been enough.

**Comment by devarajphukan** [comment]
@imotov, the delete request was hanging on for more than 8 hours. I did not know that only a restart of the current master would have solved it, so did it for the whole cluster.

**Comment by imotov** [comment]
It's possible that there are still some issues with azure plugins that causing it get stuck in waiting for the response forever. However since you have restarted the cluster, it's hard to tell what exactly was going on there. If it will happen again, please run the following command on the cluster where the snapshot is stuck and include its result with your report:
```
curl "http://localhost:9200/_nodes/hot_threads?threads=10000&ignore_idle_threads=false"
```

**Comment by devarajphukan** [comment]
@imotov the above curl request works well to stop the delete process which is stuck. This command stopped both the taking snapshot and deleting snapshot processes and the status is "FAILED". After this that particular snapshot needs to be deleted again.

**Comment by imotov** [comment]
@devarajphukan yes, the `DELETE` command is supposed to stop the current snapshot, it might take a while but it shouldn't hang for 8 hours. What I am saying is, if you will have a situation when you will have a snapshot stuck and `DELETE` command will hang for a few hours, please execute the `hot_nodes` command that I mentioned above and paste here the result. Does it make sense?

**Comment by devarajphukan** [comment]
@imotov, yes sure. Will get back with the results, when I get the same issue. Anyway, thanks for the help.

**Comment by devarajphukan** [comment]
For anyone getting the similar issue here's a cron script in python that takes snapshots and deletes older ones https://github.com/devarajphukan/ElasticSearchSnapshotting/blob/master/take_snapshot_cron_script.py

**Comment by colings86** [comment]
No further feedback. Please reopen if this issue is still reproducable

**Comment by nikoren** [comment]
I am having the same issue with 2.3, can't cancel snapshot even after rolling restart of service. 
Should I open new issue for that?
